src/macro.ts

```diff
@@ -0,0 +1,126 @@
+// const printAST = require('ast-pretty-print')
+import type babelCore from '@babel/core'
+import {createMacro} from 'babel-plugin-macros'
+import type {MacroHandler} from 'babel-plugin-macros'
+import {getReplacement} from './helpers'
+
+const prevalMacros: MacroHandler = function prevalMacros({
+  references,
+  state,
+  babel,
+}) {
+  const fileOpts = state.file.opts
+  references.default.forEach(referencePath => {
+    const path = referencePath as babelCore.NodePath<babelCore.types.Identifier>
+    const targetPath = path.parentPath
+    if (targetPath.type === 'TaggedTemplateExpression') {
+      asTag(
+        targetPath as babelCore.NodePath<babelCore.types.TaggedTemplateExpression>,
+        fileOpts,
+        babel,
+      )
+    } else if (targetPath.type === 'CallExpression') {
+      asFunction(
+        targetPath as babelCore.NodePath<babelCore.types.CallExpression>,
+        fileOpts,
+        babel,
+      )
+    } else if (targetPath.type === 'JSXOpeningElement') {
+      const jsxElement = targetPath.parentPath
+      asJSX(
+        jsxElement as babelCore.NodePath<babelCore.types.JSXOpeningElement>,
+        fileOpts,
+        babel,
+      )
+    } else if (targetPath.type === 'JSXClosingElement') {
+      // That's okay, we already prevaled this one on its opening element.
+    } else {
+      throw new Error(
+        `babel-plugin-preval/macro can only be used as tagged template expression, function call or JSX element. You tried ${referencePath.parentPath.type}.`,
+      )
+    }
+  })
+}
+
+function asTag(
+  path: babelCore.NodePath<babelCore.types.TaggedTemplateExpression>,
+  fileOpts: babelCore.TransformOptions,
+  babel: typeof babelCore,
+) {
+  const string = path.get('quasi').evaluate().value as string
+  const replacement = getReplacement({
+    string,
+    fileOpts,
+    babel,
+  })
+
+  if (!replacement) return
+  path.replaceWith(replacement)
+}
+
+function asFunction(
+  path: babelCore.NodePath<babelCore.types.CallExpression>,
+  fileOpts: babelCore.TransformOptions,
+  babel: typeof babelCore,
+) {
+  const argumentsPaths = path.get('arguments')
+  const string = argumentsPaths[0].evaluate().value as string
+  const replacement = getReplacement({
+    string,
+    fileOpts,
+    babel,
+  })
+
+  if (!replacement) return
+  argumentsPaths[0].parentPath.replaceWith(replacement)
+}
+
+// eslint-disable-next-line no-unused-vars
+function asJSX(
+  path: babelCore.NodePath<babelCore.types.JSXOpeningElement>,
+  fileOpts: babelCore.TransformOptions,
+  babel: typeof babelCore,
+) {
+  const children = path.get('children') as Array<
+    babelCore.NodePath<babelCore.types.JSXExpressionContainer>
+  >
+  // It's a shame you cannot use evaluate() with JSX
+  // @ts-expect-error value isn't on all nodes
```

Comment:
Can we not narrow the node type instead of ignoring it?

Reply:
I think `value` is on `JSXText` nodes only.
Wonder if we could use `@babel/types` and do something like
```js
  const node = children[0].node
  if (!isJSXText(node)) return
```

Reply:
I think that's a good idea

Reply:
@juhanakristian mentioned in discord that `babel-plugin-codegen` is [handling this with a `@ts-expect-error` comment as well](https://github.com/kentcdodds/babel-plugin-codegen/blob/5f893c50d941e1b397f3d389f41daad3fca389e8/src/replace.ts#L198), so we've decided to leave this for now an see what the repo owners think.